tezedge_state/src/proposer/mod.rs

```diff
@@ -0,0 +1,948 @@
+
+    fn start_listening_to_server_events(&mut self) -> io::Result<()>;
+    fn stop_listening_to_server_events(&mut self);
+
+    fn accept_connection(&mut self, event: &Self::NetworkEvent) -> Option<&mut Peer<Self::Stream>>;
+
+    /// Blocks until events are available or until timeout passes.
+    ///
+    /// New events fill the passed `events_container`, removing all
+    /// previous events from the container.
+    ///
+    /// If timeout passes, [Event::Tick] will be added to `events_container`.
+    fn wait_for_events(&mut self, events_container: &mut Self::Events, timeout: Option<Duration>);
+
+    fn get_peer(&mut self, address: &PeerAddress) -> Option<&mut Peer<Self::Stream>>;
+    fn get_peer_or_connect_mut(
+        &mut self,
+        address: &PeerAddress,
+    ) -> io::Result<&mut Peer<Self::Stream>>;
+    fn get_peer_for_event_mut(
+        &mut self,
+        event: &Self::NetworkEvent,
+    ) -> Option<&mut Peer<Self::Stream>>;
+
+    fn disconnect_peer(&mut self, peer: &PeerAddress);
+}
+
+/// Internal clock of proposer.
+#[derive(Clone)]
+struct InternalClock {
+    time: Instant,
+    elapsed: Duration,
+}
+
+impl InternalClock {
+    fn new(initial_time: Instant) -> Self {
+        Self {
+            time: initial_time,
+            elapsed: Duration::new(0, 0),
+        }
+    }
+
+    fn update(&mut self, new_time: Instant) -> &mut Self {
+        if self.time >= new_time {
+            // just to guard against panic.
+            return self;
+        }
+        self.elapsed += new_time.duration_since(self.time);
+        self.time = new_time;
+        self
+    }
+
+    /// Result of every call of this method MUST be passed to state machine
+    /// `TezedgeState` through proposal, otherwise internal clock will mess up.
+    #[inline]
+    fn take_elapsed(&mut self) -> Duration {
+        std::mem::replace(&mut self.elapsed, Duration::new(0, 0))
+    }
+}
+
+#[derive(Clone)]
+pub struct TezedgeProposerConfig {
+    pub wait_for_events_timeout: Option<Duration>,
+    pub events_limit: usize,
+    pub record: bool,
+    pub replay: bool,
+}
+
+/// Tezedge proposer wihtout `events`.
+#[derive(Clone)]
+struct TezedgeProposerInner<Efs, M> {
+    config: TezedgeProposerConfig,
+    requests: Vec<TezedgeRequest>,
+    notifications: Vec<Notification>,
+    effects: Efs,
+    time: InternalClock,
+    proposal_persister: Option<ProposalPersisterHandle>,
+    /// Proposal loader for replay.
+    proposal_loader: Option<ProposalLoader>,
+    state: TezedgeStateWrapper,
+    manager: M,
+}
+
+impl<S, NetE, Es, Efs, M> TezedgeProposerInner<Efs, M>
+where
+    S: Read + Write,
+    NetE: NetworkEvent + Debug,
+    Efs: Effects + Debug,
+    M: Manager<Stream = S, NetworkEvent = NetE, Events = Es>,
+{
+    #[inline]
+    fn wait_for_events(&mut self, events: &mut Es) {
+        self.manager
+            .wait_for_events(events, self.config.wait_for_events_timeout)
+    }
+
+    // fn handle_event(&mut self, event: Event<NetE>) {
```

Comment:
@binier 
do we need this commented code?

Reply:
It's not used at the moment and just was tired of keeping that impl in sync with `handle_event_ref`. Basically it was for more flexibility, so that `NetworkEvent` can be owned during testing, instead of being a ref (like it is with mio). Maybe it's better to keep it there so that if we want to handle owned events for some test, we can uncomment that.

Reply:
Nvm, removed.